subsys/net/lib/dns/mdns_responder.c

```diff
@@ -1481,21 +1472,275 @@ static int init_listener(void)
 }
 
 #if defined(CONFIG_MDNS_RESPONDER_PROBE)
+
+#define ANNOUNCE_TIMEOUT 1 /* in seconds, RFC 6762 ch 8.3 */
+
+/* ATM we do only .local announcing, SD announcing is TODO */
+
+static int send_unsolicited_response(struct net_if *iface,
+				     int sock,
+				     sa_family_t family,
+				     struct sockaddr *src_addr,
+				     size_t addrlen,
+				     struct net_buf *answer)
+{
+	socklen_t dst_len;
+	int ret;
+
+	COND_CODE_1(IS_ENABLED(CONFIG_NET_IPV6),
+		    (struct sockaddr_in6), (struct sockaddr_in)) dst;
+
+	ret = setup_dst_addr(sock, family, (struct sockaddr *)&dst, &dst_len);
+	if (ret < 0) {
+		NET_DBG("unable to set up the response address");
+		return ret;
+	}
+
+	ret = zsock_sendto(sock, answer->data, answer->len, 0,
+			   (struct sockaddr *)&dst, dst_len);
+	if (ret < 0) {
+		ret = -errno;
+	} else {
+		net_stats_update_dns_sent(iface);
+	}
+
+	return ret;
+}
+
+static struct net_buf *create_unsolicited_mdns_answer(struct net_if *iface,
+						      const char *name,
+						      uint32_t ttl,
+						      struct mdns_monitor_iface_addr *addr_list,
+						      size_t addr_list_len)
+{
+	struct net_buf *answer;
+	uint16_t answer_count;
+	int len;
+
+	answer = net_buf_alloc(&mdns_msg_pool, BUF_ALLOC_TIMEOUT);
+	if (answer == NULL) {
+		return NULL;
+	}
+
+	setup_dns_hdr(answer->data, 1);
+	net_buf_add(answer, DNS_MSG_HEADER_SIZE);
+
+	answer_count = 0U;
+
+	for (size_t i = 0; i < addr_list_len; i++) {
+		uint16_t type;
+		size_t left;
+
+		if (!addr_list[i].in_use) {
+			continue;
+		}
+
+		if (iface != addr_list[i].iface) {
+			continue;
+		}
+
+		if (!addr_list[i].updated) {
+			continue;
+		}
+
+		if (!addr_list[i].added) {
+			continue;
+		}
+
+		if (addr_list[i].addr.family == AF_INET) {
+			type = DNS_RR_TYPE_A;
+		} else if (addr_list[i].addr.family == AF_INET6) {
+			type = DNS_RR_TYPE_AAAA;
+		} else {
+			NET_DBG("Unknown family %d", addr_list[i].addr.family);
+
+			net_buf_unref(answer);
+			return NULL;
+		}
+
+		if (answer_count == 0) {
+			len = strlen(name);
+		}
+
+		left = net_buf_tailroom(answer);
+		if ((answer_count == 0 &&
+		     left < (1 + len + 1 + 5 + 1 + 2 + 2 + 4 + 4 +
+			     (type == DNS_RR_TYPE_A ? 4 : 16))) ||
+		    (answer_count > 0 &&
+		     left < (1 + 1 + 2 + 2 + 4 + 4 +
+			     (type == DNS_RR_TYPE_A ? 4 : 16)))) {
+			NET_DBG("No more space (%u left)", left);
+			net_buf_unref(answer);
+			return NULL;
+		}
+
+		if (answer_count == 0) {
+			net_buf_add_u8(answer, len);
+			net_buf_add_mem(answer, name, len);
+			net_buf_add_u8(answer, sizeof(".local") - 2);
+			net_buf_add_mem(answer, &".local"[1], sizeof(".local") - 2);
+			net_buf_add_u8(answer, 0U);
+		} else {
+			/* Add pointer to the name (compression) */
+			net_buf_add_u8(answer, 0xc0);
+			net_buf_add_u8(answer, 0x0c);
+		}
+
+		net_buf_add_be16(answer, type);
+		net_buf_add_be16(answer, DNS_CLASS_IN);
+		net_buf_add_be32(answer, ttl);
+
+		if (type == DNS_RR_TYPE_A) {
+			net_buf_add_be16(answer, 4);
```

Comment:
Maybe use `sizeof(struct in_addr)`/`sizeof(struct in6_addr)` instead of magic numbers?

Reply:
Yes, will do.